feat: Frontend controlled map style

marius-at-atb opened the pull request:
Closes https://github.com/AtB-AS/kundevendt/issues/19557
Closes https://github.com/AtB-AS/kundevendt/issues/10821
Closes https://github.com/AtB-AS/kundevendt/issues/19451
Closes https://github.com/AtB-AS/kundevendt/issues/19259

First of all excuse the large PR, but it is kinda needed to avoid incomplete/invalid steps along the way 😅 

Feel free to ask me for closer explanations if needed!

What's happening in this PR:

- The `Map` component has been updated. That means only the map tab and choose location maps are updated for now.
- New hook: `useMapboxJsonStyle`, which uses the files inside `mapbox-styles` to get the style locally, but sprites from the url defined in remote config. This only shows geographical features, not NSR items.
- New component: `NationalStopRegistryFeatures`. This is where the NSR items are shown now. In order to access the data, new mapbox env variables have been added.
- New hooks for determining the style for icons and text for NSR and Shared Mobility items.
- No longer zooming in on a path when selecting a feature.
- Mobility components have been updated to use the new icons. These components will be removed when using the tile server, so let's not spend too much time on them..
- Dark mode support.

How the zoom logic works:
- In nsr-layers.ts, showAsDefaultAtZoomLevel is defined for each NSR layer. 3 zoom levels before, the minimized version will transition from small and invisible, to full size and visible. Then when `showAsDefaultAtZoomLevel` is reached, the icon switches to default. The label will then start to fade in as well.

Before:
<img width="200" src="https://github.com/user-attachments/assets/c23ae29c-b836-486c-a68d-de06ce76e8ab" />

After:
<section>
<img width="200" src="https://github.com/user-attachments/assets/40ff2aa7-9548-40e0-8a78-234ae3558907" />
<img width="200" src="https://github.com/user-attachments/assets/a65f67d3-efb3-431f-9a0b-773213784029" />
<img width="200" src="https://github.com/user-attachments/assets/5b77c219-f1eb-4c5e-9ad2-91b440369341" />
<img width="200" src="https://github.com/user-attachments/assets/5dc7a84a-6ce5-497e-81c2-f4ec792351f2" />
<img width="200" src="https://github.com/user-attachments/assets/5b3ac3f7-cb32-4037-b81d-8f6175340668" />
<img width="200" src="https://github.com/user-attachments/assets/ecfbb513-e66f-43eb-bdba-8e9092d477cb" />
</section>

Acceptance Criteria:
- [ ] Light theme in light mode
- [ ] Dark theme in dark mode
- [ ] 200% zoom
- [ ] Performance - it should not be much slower than before
- [ ] When selecting a map item, it should
      - [ ] show a custom pin to indicate the selected element
      - [ ] other items should become minimized (apart from scooter clusters, that will be addressed later)
- [ ] When zooming in, the stop places should
      - [ ] first be invisible and tiny in the minimized state
      - [ ] then transition to reach full size and have no transparency in the minimized state as you zoom in
      - [ ] switch to the default icon
      - [ ] the label should then start becoming visible
      - [ ] The quays should start small, and when full size is reached, the text should start fading in
      - NOTE: different types of stop places are configured to reach these states at different zoom levels
- [ ] Vehicles and stations should still work with clustering and show correct numbers (navigate to Oslo in the map to test this)

marius-at-atb commented on src/components/map/Map.tsx:
```diff
@@ -184,6 +191,15 @@ export const Map = (props: MapProps) => {
     async (feature: Feature) => {
       if (!isFeaturePoint(feature)) return;
 
+      // should split components instead of this, ExploreLocation should only depend on location state, not features
```
Upcoming PR..

marius-at-atb replied:
https://github.com/AtB-AS/mittatb-app/pull/5076

marius-at-atb commented on src/components/map/Map.tsx:
```diff
@@ -252,7 +269,12 @@ export const Map = (props: MapProps) => {
-          {...MapViewConfig}
+          {...{
+            ...MapViewConfig,
+            // only updating Map.tsx for now.
```
Upcoming PR for the other maps

marius-at-atb replied:
https://github.com/AtB-AS/mittatb-app/pull/5077

marius-at-atb commented on src/components/map/Map.tsx:
```diff
@@ -275,12 +297,22 @@ export const Map = (props: MapProps) => {
+          />
+
+          {props.selectionMode !== 'ExploreLocation' && (
+            <SelectedFeatureIcon selectedFeature={selectedFeature} />
```
Keeping SelectedFeatureIcon as a separate icon. When using the tile server (upcoming PR), this is needed to keep a visible item selected when zooming out.

marius-at-atb commented on src/components/map/components/mobility/Bicycles.tsx:
```diff
@@ -1,91 +1,102 @@
 import React, {RefObject, useRef} from 'react';
```
All files inside `mobility/` will be deleted when switching to use the tile server, so no need to perfect these files.

marius-at-atb commented on src/components/map/mapbox-styles/pin-types.ts:
```diff
@@ -0,0 +1,49 @@
+import {Theme, Themes} from '@atb-as/theme';
```
Note: Types and names of the icons will be iterated on in a follow-up PR.

jorelosorio commented on src/components/map/Map.tsx:
```diff
@@ -162,14 +166,17 @@ export const Map = (props: MapProps) => {
    * overhead.
    */
   const onMapIdle = (state: MapState) => {
+    console.log('onMapIdle');
```
There are some console logs here that print some warnings.

jorelosorio commented on src/components/map/Map.tsx:
```diff
@@ -162,14 +166,17 @@ export const Map = (props: MapProps) => {
-    setMapRegion((prevMapRegion) =>
-      isEqual(prevMapRegion, newMapRegion) ? prevMapRegion : newMapRegion,
-    );
+    if (!isEqual(mapRegion, newMapRegion)) {
```
Maybe a stateless var instead. But perhaps is needed to be in that way.

marius-at-atb replied:
I like this suggestion!
It is possible to change the mapRegion state to a ref instead, and update vehicles and stations directly instead of in a useEffect:

```ts
  const mapRegionRef = useRef<MapRegion>();
  const updateVehiclesAndStations = useCallback(
    (mapRegion: MapRegion) => {
      if (!mapRegion) return;
      mapRegionRef.current = mapRegion;
      updateRegionForVehicles?.(mapRegion);
      updateRegionForStations?.(mapRegion);
    },
    [updateRegionForVehicles, updateRegionForStations],
  );
```
and then instead of setMapRegion:
```ts
    if (!isEqual(mapRegionRef.current, newMapRegion)) {
      updateVehiclesAndStations(newMapRegion);
    }
```
Also for `onDidFinishLoadingMap`:
```ts
    updateVehiclesAndStations({
      visibleBounds,
      zoomLevel,
      center,
    });
```

Seems to work well when testing, so I actually prefer this.
However it is another change, so not sure if it should be included in this PR 🤔

marius-at-atb replied:
Update: MapV2 using VectorSource instead now, no longer using onMapIdle/mapRegion.

jorelosorio commented on src/components/map/components/SelectedFeatureIcon.tsx:
```diff
@@ -0,0 +1,78 @@
+}: {
+  selectedFeature?: Feature<Point, GeoJsonProperties>;
+}) => {
+  const pinType: PinType =
```
I suggest moving this ternary to a switch can be easier to read.

marius-at-atb replied:
https://github.com/AtB-AS/mittatb-app/pull/4937/commits/3c99ebe42a65f90397ed234e8e02335d733c678e

jorelosorio commented on src/components/map/components/SelectedFeatureIcon.tsx:
```diff
@@ -0,0 +1,78 @@
+  const {iconStyle, textStyle} = useMapSymbolStyles(
+    selectedFeature?.properties?.id,
+    pinType,
+    1.61, // increased text size since the icon is larger
```
How does it work if font size increases?

marius-at-atb replied:
This number multiplies the text size. This is needed here since the selected version of the icons is larger.

jorelosorio commented on src/components/map/components/national-stop-registry-features/nsr-utils.ts:
```diff
@@ -0,0 +1,163 @@
+        ['exponential', 1.5],
+        ['zoom'],
+        14,
+        textSizeFactor * 13,
```
Is not very intuitive why to multiply by 13 here or 27 below, would be nice to add some comments about it.

marius-at-atb replied:
Comment added: https://github.com/AtB-AS/mittatb-app/pull/4937/commits/2a0c6a034d6e40b151bda09fb628c1c5b651d6cd

jorelosorio commented on src/components/map/components/national-stop-registry-features/use-nsr-circle-layers.ts:
```diff
@@ -0,0 +1,49 @@
+        circleOpacity,
+
+        circleColor: theme.color.transport.city.primary.background,
+        circleStrokeColor: themeName === 'light' ? '#ffffff' : '#000000',
```
Can this be set in the theme?

marius-at-atb replied:
We don't have shadow specs in design system from before. I guess we should at some point though.

jorelosorio commented on src/components/map/components/national-stop-registry-features/use-nsr-symbol-layers.ts:
```diff
@@ -0,0 +1,113 @@
+          textHaloColor,
+          textVariableAnchor,
+          textAnchor: 'top',
+          textFont: ['DIN Offc Pro Regular', 'Arial Unicode MS Regular'],
```
Not sure if those are default fonts, we might use defaults ones.

marius-at-atb replied:
These are the same as in the map style currently in prod, so it will be the same as before.

jorelosorio commented on src/components/map/hooks/use-map-symbol-styles.ts:
```diff
@@ -0,0 +1,185 @@
+    'generic',
+    [
+      'case',
+      ['==', ['slice', systemId, 0, 4], 'ryde'],
```
can those operators be taken for a dynamic list of operators, maybe later we want to turn some of them off.

jorelosorio commented:
In general, the code looks good, nice work. There are some small comments here and there, in general, I see that there are a lot of values, offsets and fonts set in the map, some of them look a bit random without more context, and it might be difficult to change to another way but would be nice to have some comments, in general, the documentation of the code looks good.

gorandalum reviewed (commented):
In general I think this looks good! The map code was pretty complex from before, but from what I can see there isn't too much increase in complexity from this. 

Also the map looks pretty nice, and I couldn't find any glaring bugs when trying it out. One thing I noticed was that the mobility bottom sheet was all the way up to the icon if the max height was used. The bus and mobility bottom sheets should probably have the same max height, and the icon placement should be set based on the shared max height.

When removing a feature (like the map lines) it would be nice with a reference to the decision and background/insight for it. When we did insight into the map originally the user tests showed that it was a requested feature. The reason I remember it is because it was pretty complex to implement with the correct zooming animation, and I tried to convince the designers at the time to skip it for the initial map release, with no luck 😅 

All in all a thumbs up from me! 👍 I won't approve it since my review was pretty rough, and it should still be pair-reviewed by you and Johannes.

tormoseng commented:
Observations (listing as I go):
1. Seems like the speed boat icon hit range is to wide in some cases. In this example it's not possible to open the train or bus station (screenshot_1)
2. When visible the bus icon should be "prioritized" on top of other transport icons (screenshot_2)

<details>
<summary>screenshot_1</summary>

![IMG_5341](https://github.com/user-attachments/assets/963a4934-d748-4ef2-b6b7-11022212c5f9)

</details>

<details>
<summary>screenshot_2</summary>

![IMG_5340](https://github.com/user-attachments/assets/0c386b45-3abc-44d3-a9c2-2a7645f4ab81)

</details>

rosvik commented on src/travel-details-map-screen/TravelDetailsMapScreenComponent.tsx:
```diff
@@ -210,7 +210,11 @@ export const TravelDetailsMapScreenComponent = ({
         {stations && (
-          <Stations stations={stations.stations} mapCameraRef={mapCameraRef} />
+          <Stations
+            selectedFeaturePropertyId="" // no active selection feature used here
```
Should this prop be optional?

rosvik commented on src/components/map/utils.ts:
```diff
@@ -223,8 +226,8 @@ export const getVisibleRange = (visibleBounds: Position[]) => {
-  isStation(entityFeature) || isStopPlace(entityFeature);
+// disable for now
+export const shouldShowMapLines = (_entityFeature: Feature<Point>) => false; //isStationFeature(entityFeature) || isStopPlaceFeature(entityFeature);
 
-export const shouldZoomToFeature = (entityFeature: Feature<Point>) =>
-  isStation(entityFeature) || isStopPlace(entityFeature);
+// disable for now
+export const shouldZoomToFeature = (_entityFeature: Feature<Point>) => false; //isStationFeature(entityFeature) || isStopPlaceFeature(entityFeature);
```
I assume this will change soon™, but I think it's better to remove this and related logic to keep the code clean on master.

marius-at-atb replied:
In case it ever becomes relevant to re-introduce something similar, I think it is nice to have an isolated PR for it.
Also this PR already has a lot going on, so I'll make an immediate follow-up PR with this :+1:

rosvik commented on src/components/map/mapbox-styles/README.md:
Nice 🙌

marius-at-atb commented on src/components/map/mapbox-styles/pin-types.ts:
```diff
@@ -0,0 +1,49 @@
+
+export type StationIconCode = 'sharedcar' | 'citybike';
+
+export type PinScooterCompany = 'generic' | 'voi' | 'ryde' | 'tier'; // todo: change 'tier' to 'dott' when sprite ready
```
This is used in order to show correct colors per company. I suppose we could use color names instead of company names.
<img width="200" src="https://github.com/user-attachments/assets/2fff5687-3817-4a5f-b343-8f202d258b68" />

marius-at-atb commented:
@rosvik Good catch with the status bar color! Updated in https://github.com/AtB-AS/mittatb-app/pull/4937/commits/af01eebf04d948f0f650febf4e3907690571043a. Let me know if I removed too much here.

marius-at-atb commented:
@tormoseng 
1. Will look into this a bit.
2. Updated [here](https://github.com/AtB-AS/mittatb-app/pull/4937/commits/e1b2f6df318979bd94c7904420fceccfda310b11) so that the last listed item in nsrSymbolLayers is on top. Not 100% sure, but I think this is now the same order as in prod from before.
3. We can look into this.
4 and 5: Not a part of changes in this PR I think, but should be looked at :+1:
6: This has been addressed with onPress events directly on sources so that we don't need to wait for the querying of features where you clicked, and with useCallbacks/useMemos to minimize the number of renders of the Map component. However, it is still a bit slower than before because it supports more features. If still too slow, we probably go to the tile server directly.

rosvik commented on ios/Podfile.lock:
```diff
@@ -3434,13 +3434,13 @@ SPEC CHECKSUMS:
   AbtProtobufTokenIosStubs: 76426ea8c7c8cf76a750b04bcd4be8288bafae70
   ApiLogiOSLib: bda60ece3f53fcbc6d1cfcbba4c0993057e6c2a7
-  BarcodeiOSLib: db5edae8312bf90e30673fbfe00832e2e3ea2315
+  BarcodeiOSLib: 45c0bb7f5f709c1c4147db333ee532169475e464
```
Probably not. Doesn't look like the dependencies actually has changed, but rather an effect of different xcode versions.

marius-at-atb replied:
Reverted

reidzeibel reviewed (approved):
I think with the feature toggle, this PR can be merged as-is, and we can fix the issues and some nitpicks on future PRs.

rosvik commented on src/travel-details-map-screen/TravelDetailsMapScreenComponent.tsx:
```diff
@@ -33,15 +32,15 @@ import {DirectionArrow} from './components/DirectionArrow';
 } from '@rnmapbox/maps/lib/typescript/src/components/MapView';
 
-const EMPTY_FILTER = {};
+// todo: re-add this
+// import {VehiclesAndStations} from '@atb/components/map/components/mobility/VehiclesAndStations';
```
```suggestion
```

rosvik commented on yarn.lock:
Maybe fix the line break, so that this won't trigger uncached builds.

tormoseng commented:
✅  Looking good! Will continue test the map, and note any observations in https://github.com/AtB-AS/kundevendt/issues/20260
